### PR TITLE
Create Macro for Implementing `SlotRange`

### PR DESCRIPTION
**Pull request by shawntabrizi**
This PR sets up the work needed to increase the number of lease periods per slot to 8 (or any number).

That will come in a future PR if we determine it is what we want to do.

### CODE REVIEW

File: runtime/common/Cargo.toml [hunk 1]
```diff
@@ -13,6 +13,7 @@ rustc-hex = { version = "2.1.0", default-features = false }
 serde = { version = "1.0.123", default-features = false }
 serde_derive = { version = "1.0.117", optional = true }
 static_assertions = "1.1.0"
+paste = "1.0"
```
Comment (shawntabrizi) [inline runtime/common/Cargo.toml:16]:
```suggestion
```

File: runtime/common/slot_range_helper/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	) => {
+		$crate::generate_slot_range_enum! {
+			/// A compactly represented sub-range from the series.
+			#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, $crate::Encode, $crate::Decode)]
+			#[repr(u8)]
+			pub enum SlotRange { $( $parsed ),* }
+		}
+	};
+	// A trick from StackOverflow to generate the final Enum object.
+	($i:item) => { $i };
```
Comment (thiolliere) [inline runtime/common/slot_range_helper/src/lib.rs:89]: I don't understand why you call recursively
```suggestion
		/// A compactly represented sub-range from the series.
		#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, $crate::Encode, $crate::Decode)]
		#[repr(u8)]
		pub enum SlotRange { $( $parsed ),* }
	};
```

File: runtime/common/slot_range_helper/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+#[macro_export]
```
Comment (thiolliere) [inline runtime/common/slot_range_helper/src/lib.rs:92]:
```suggestion
#[macro_export]
#[doc(hidden)]
```

File: runtime/common/slot_range_helper/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+		}
+	};
+}
+
+#[macro_export]
+macro_rules! generate_slot_range_count {
+	(@
+		$( $idents:ident )*
+	) => {
+		#[allow(dead_code, non_camel_case_types)]
```
Comment (thiolliere) [inline runtime/common/slot_range_helper/src/lib.rs:163]:
```suggestion
		#[doc(hidden)]
		#[allow(dead_code, non_camel_case_types)]
```

File: runtime/common/slot_range_helper/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+				.map_err(|_| "start too far")?;
+			match (first, first + count) {
+				$( ($t1, $t2) => { Ok(SlotRange::$parsed) })*
+				_ => Err("bad range"),
+			}
+		}
+	};
+}
+
+#[macro_export]
```
Comment (thiolliere) [inline runtime/common/slot_range_helper/src/lib.rs:158]:
```suggestion
#[macro_export]
#[doc(hidden)]
```

File: runtime/common/slot_range_helper/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+		$( $parsed:ident ( $t1:expr, $t2:expr ) )*
+	) => {
+		pub fn new_bounded<
+			Index: $crate::Add<Output=Index> + $crate::CheckedSub + Copy + Ord + From<u32> + $crate::TryInto<u32>
+		>(
+			initial: Index,
+			first: Index,
+			last: Index
+		) -> $crate::result::Result<Self, &'static str> {
+			if first > last || first < initial || last > initial + 3.into() {
```
Comment (thiolliere) [inline runtime/common/slot_range_helper/src/lib.rs:139]: what is the 3 here ? should it be `SLOT_RANGE_COUNT - 1` ?

File: runtime/common/slot_range_helper/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+		pub fn as_pair(&self) -> (u8, u8) {
+			match self {
+				$( SlotRange::$parsed => { ($t1, $t2) } )*
+			}
+		}
+	};
+}
+
+#[macro_export]
```
Comment (thiolliere) [inline runtime/common/slot_range_helper/src/lib.rs:112]:
```suggestion
#[macro_export]
#[doc(hidden)]
```

File: runtime/common/slot_range_helper/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+			pub fn len(&self) -> usize {
+			match self {
+				// len (0, 2) = 2 - 0 + 1 = 3
+				$( SlotRange::$parsed => { ( $t2 - $t1 + 1) } )*
+			}
+		}
+	};
+}
+
+#[macro_export]
```
Comment (thiolliere) [inline runtime/common/slot_range_helper/src/lib.rs:127]:
```suggestion
#[macro_export]
#[doc(hidden)]
```

**Review by thiolliere — COMMENTED** [review]
it will need doc and test, but implementation is good to me, except https://github.com/paritytech/polkadot/pull/2788#discussion_r605454629

File: runtime/common/src/slot_range.rs [hunk 9]
```diff
@@ -17,139 +17,7 @@
...
-			6 => SlotRange::OneThree,
-			7 => SlotRange::TwoTwo,
-			8 => SlotRange::TwoThree,
-			9 => SlotRange::ThreeThree,
-			_ => return Err(()),
-		})
-	}
-}
+// An alternative implementation supporting 8 lease periods per slot.
+// slot_range_helper::generate_slot_range!(Zero(0), One(1), Two(2), Three(3), Four(4), Five(5), Six(6), Seven(7));
```
[the remark this answers is not in the document]
Comment (shawntabrizi) [reply runtime/common/src/slot_range.rs:23]: We might switch to it in the near future, so I just left it in

[the remark this answers is not in the document]
Comment (bkchr) [reply runtime/common/src/slot_range.rs:23]: And if we not switch to it, we keep dead code :P 

Writing this code also takes 10 seconds :P

File: runtime/common/slot_range_helper/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+#[doc(hidden)]
+macro_rules! generate_lease_period_per_slot {
+	(@
+		$( $idents:ident )*
+	) => {
+		#[doc(hidden)]
+		#[allow(dead_code, non_camel_case_types)]
+		enum __LeasePeriodPerSlotTemp { $($idents,)* __LeasePeriodPerSlotLast }
+		const LEASE_PERIODS_PER_SLOT: usize = __LeasePeriodPerSlotTemp::__LeasePeriodPerSlotLast as usize;
+	};
```
Comment (bkchr) [inline runtime/common/slot_range_helper/src/lib.rs:215]:
```suggestion
	(
		$start:ident $( $rest:ident )*
	) => {
		crate::generate_lease_period_per_slot!(@inner 1; $( $rest )*);
	};
	(@inner
		$count:expr;
		$start:ident $( $rest:ident )*
	) => {
		crate::generate_lease_period_per_slot!(@inner $count + 1; $( $rest )*);
	};
	(@inner
		$count:expr;
	) => {
		const LEASE_PERIODS_PER_SLOT: usize = $count;
	};
```

Comment (bkchr) [reply runtime/common/slot_range_helper/src/lib.rs:215]: Less hacky :P

Comment (bkchr) [reply runtime/common/slot_range_helper/src/lib.rs:215]: The other macro could do the same

File: runtime/common/slot_range_helper/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+macro_rules! generate_slot_range_as_pair {
+	(@
+		$( $parsed:ident ( $t1:expr, $t2:expr ) )*
+	) => {
+		/// Return true if two `SlotRange` intersect in their lease periods.
+		pub fn intersects(&self, other: SlotRange) -> bool {
+			let a = self.as_pair();
+			let b = other.as_pair();
+			b.0 <= a.1 && a.0 <= b.1
+			// == !(b.0 > a.1 || a.0 > b.1)
```
[the remark this answers is not in the document]
Comment (shawntabrizi) [reply runtime/common/slot_range_helper/src/lib.rs:123]: This is actually a "code comment". I think nice to keep in.

[the remark this answers is not in the document]
Comment (bkchr) [reply runtime/common/slot_range_helper/src/lib.rs:126]: Ahh okay ;)

**Review by bkchr — APPROVED** [review]

**Review by gavofyork — APPROVED** [review]

**Pull request merged**